dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+
+    StageOut *out = ctx.output.data[0];
+    int N = stage.num_samples();
+    for (int i = 0; i < N; i++) {
+      assert(stage.shape[i].inner == stage.shape[0].inner);
+      assert(stage.shape[i].outer == stage.shape[0].outer);
+      assert(stage.shape[i].reduced_in == 1);
+      assert(stage.shape[i].reduced_out == (i ? 0 : 1));
+    }
+
+    int64_t sample_size = stage.shape[0].input_elements();
+
+    dim3 block(1024);
+    dim3 grid(std::min<int>(div_ceil(sample_size, 1024), 1024));
+
+    auto *pre = This().template GetPreprocessorBanks<1>(wa, -1, bool_const<is_first>());
+    auto post = This().GetPostprocessor(bool_const<is_last>());
+
+    wa.CopyParamsToDevice(ctx.stream);
+
+    auto *gpu_in  = wa.GetDeviceParam(in);
+    auto *gpu_pre = wa.GetDeviceParam(pre);
+
+    ReduceSamplesKernel<<<grid, block, 0, ctx.stream>>>(
+      out, gpu_in, sample_size, N, This().GetReduction(), gpu_pre, post);
+
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  template <typename StageOut, typename StageIn, bool is_first, bool is_last>
+  void LaunchStage(Context &ctx, const ReductionStage &stage,
+                   ReductionKindTag<ReductionKind::None>) {
+    assert(is_last);
+    assert(!stage.shape.empty());
+    WorkArea &wa = ctx.work_area;
+
+    const StageIn *const *in = InputPtrs<StageIn>(ctx, stage);
+    StageOut *const *out = OutputPtrs<StageOut>(ctx, stage);
+
+    int N = stage.num_samples();
+    int64_t *sample_sizes = wa.ParamBuffer<int64_t>(N);
+
+    for (int i = 0; i < N; i++) {
+      sample_sizes[i] = stage.shape[i].input_elements();
+      assert(stage.shape[i].input_elements() == stage.shape[0].output_elements());
+    }
+
+    dim3 block(1024);
+    dim3 grid(std::max(div_ceil(1024, N), 32), N);
+
+    auto *pre = This().template GetPreprocessorBanks<1>(wa, -1, bool_const<is_first>());
+    auto *post = This().GetPostprocessors(wa, bool_const<is_last>());
+
+    wa.CopyParamsToDevice(ctx.stream);
+
+    auto *gpu_in              = wa.GetDeviceParam(in);
+    auto *gpu_out             = wa.GetDeviceParam(out);
+    const int64_t *gpu_sizes  = wa.GetDeviceParam(sample_sizes);
+    auto *gpu_pre             = wa.GetDeviceParam(pre);
+    auto *gpu_post            = wa.GetDeviceParam(post);
+
+    ReduceNoneKernel<<<grid, block, 0, ctx.stream>>>(
+      gpu_out, gpu_in, gpu_sizes, gpu_pre, gpu_post);
+
+    CUDA_CALL(cudaGetLastError());
+  }
+
+
+  void Run(KernelContext &kctx, const OutListGPU<Out> &out, const InListGPU<In> &in) {
+    assert(!stages_.empty());
+    Context ctx;
+    auto host_mem_size = buffer_sizes_.param_buffers;
+    auto gpu_mem_size = buffer_sizes_.param_buffers + buffer_sizes_.io_buffers;
+
+    ctx.stream = kctx.gpu.stream;
+    ctx.work_area.buffer_sizes = buffer_sizes_;
+    ctx.work_area.host_memory = kctx.scratchpad->Allocate<char>(AllocType::Host, host_mem_size, 64);
+    ctx.work_area.gpu_memory = kctx.scratchpad->Allocate<char>(AllocType::GPU, gpu_mem_size, 64);
```

Comment:
You're not using the same alloc&pad approach to gpu param & io buffers (in setup you configure them in two separate steps). It probably won't matter, but that's probably the only issue I will be able to notice, as the CalculateTempBuffers is incomprehensible for me rn (at least till I understand the memory use of the whole kernel).

Reply:
Right you are. Actually, I should align param_buffer's size and allocate - and reserve - all gpu memory in one go.

Reply:
Done (param_buffers are aligned in CalculateTempBuffers.